### PR TITLE
Shows the usage of transport transaction

### PR DESCRIPTION
**Pull request by danielmarbach**
Shows how transport transaction was meant to be used.

@Particular/azure-service-bus-maintainers

### CODE REVIEW

**Comment by yvesgoeleven** [comment]
The test When_dispatching_messages_in_receive_context.Should_dispatch_message_in_receive_context appears to be hanging now

**Comment by danielmarbach** [comment]
@yvesgoeleven pushed an update

**Comment by SeanFeldman** [comment]
@danielmarbach is everything committed? `When_dispatching_messages_in_receive_context` hangs locally (on my sandboxed namespace)

File: src/Transport/Seam/MessagePump.cs [hunk 1]
```diff
@@ -42,12 +42,7 @@ public Task Init(Func<PushContext, Task> pump, CriticalError criticalError, Push
 
                 circuitBreaker.Success();
 
-                var context = new ContextBag();
-
-                context.Set(receiveContext);
-
-                //todo, figure out what the TransportTransaction parameter is about
-                return messagePump(new PushContext(incoming.MessageId, incoming.Headers, incoming.BodyStream, new NoTransaction(), tokenSource, context));
+                return messagePump(new PushContext(incoming.MessageId, incoming.Headers, incoming.BodyStream, new ReceiveContextAwareTransaction(receiveContext), tokenSource, new ContextBag()));
```
Comment (SeanFeldman) [inline src/Transport/Seam/MessagePump.cs]: There's a some sort of mixup here... the `ContextBag` is empty, the receive context is sent via transaction. So the [test](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/7032515acfa946a11dbb869564ebcf287162c115/src/Tests/Seam/When_dispatching_messages_in_receive_context.cs#L65) will throw an exception and abandon the message, causing an infinite loop.

File: src/Tests/Seam/When_dispatching_messages_in_receive_context.cs [hunk 2]
```diff
@@ -62,7 +62,7 @@ public class When_dispatching_messages_in_receive_context
                 var bytes = Encoding.UTF8.GetBytes("Whatever");
                 var outgoingMessage = new OutgoingMessage("Id-1", new Dictionary<string, string>(), bytes);
 
-                var ctx = context.Context.Get<ReceiveContext>();
+                var ctx = context.Context.Get<ReceiveContextAwareTransaction>().ReceiveContext;
```
Comment (SeanFeldman) [inline src/Tests/Seam/When_dispatching_messages_in_receive_context.cs]: The `ReceiveContext` is not longer in the `ContextBag`, but in the `TransportTransaction`.

File: src/Transport/Seam/Dispatcher.cs [hunk 3]
```diff
@@ -20,24 +20,24 @@ public Dispatcher(IRouteOutgoingBatches routeOutgoingBatches, IBatcher batcher)
 
         public Task Dispatch(TransportOperations operations, ContextBag context)
         {
-            ReceiveContext receiveContext;
+            ReceiveContextAwareTransaction transaction;
             var outgoingBatches = batcher.ToBatches(operations);
 
-            context.TryGet(out receiveContext);
-            if (receiveContext == null) // not in a receive context, so send out immediately
+            context.TryGet(out transaction);
```
Comment (SeanFeldman) [inline src/Transport/Seam/Dispatcher.cs]: This is wrong. `ContextBag` is no longer set with the `ReceiveContext`.

**Comment by SeanFeldman** [comment]
@danielmarbach `IDispatchMessages.Dispatch() receives `TransportOperations` and `ContextBag`. The `ReceiveContext` is no longer set in the `ContextBag, but instead is set on the Transport operation. My quick fix is to set `ReceiveContext` on both, but that smells. Should `IDispatchMessages` receive tranport transaction rather than a context bag? I don't follow what the context bag should do here...

**Comment by danielmarbach** [comment]
@Particular/azure-service-bus-maintainers green and ready to be merged

**Pull request merged**